### ISSUE TITLE
Performance of Simple Use Case -  PyTorch vs Plain Python with Numpy

### ISSUE THREAD

**Issue by makeyourownneuralnetwork** [pytorch/pytorch#1630]
**Summary - PyTorch seems to be half as fast as plain Python for very simple networks.**

### Context
I am learning PyTorch because it promises convenience and easy access to GPU acceleration.

In learning, I re-implemented my previous very simple tutorial 3-layer neural network created from scratch in Python with Numpy. 

### Code
The two are here:

- simple python+numpy: [https://github.com/makeyourownneuralnetwork/makeyourownneuralnetwork/blob/master/part2_neural_network_mnist_data.ipynb](https://github.com/makeyourownneuralnetwork/makeyourownneuralnetwork/blob/master/part2_neural_network_mnist_data.ipynb)
- PyTorch: [https://github.com/makeyourownneuralnetwork/pytorch/blob/master/pytorch_neural_network_mnist_data.ipynb](https://github.com/makeyourownneuralnetwork/pytorch/blob/master/pytorch_neural_network_mnist_data.ipynb)

### Results
They are built to be as similar as possible - same size, same loss function, same training data, etc

The result of 2 timing tests show that the **plain python version is about twice as fast as the PyTorch version**:

- home-made simple pure python - 4**40 seconds, 458 seconds**
- simple PyTorch version - **841 seconds, 834 seconds**

### More Detail
Here's a blog which details the experiment conditions in more detail (near the bottom of the post). 

- blog  [http://makeyourownneuralnetwork.blogspot.co.uk/2017/05/learning-mnist-with-gpu-acceleration.html](http://makeyourownneuralnetwork.blogspot.co.uk/2017/05/learning-mnist-with-gpu-acceleration.html)

### What Am I Doing Wrong?
Or does PyTorch have too much overhead for simple scenarios, and only really shines for larger or GPU accelerated cases?

**Comment by pavanky** [comment]
1) Just being pedantic, you are not using plain python. You are using numpy.
2) It'll also be helpful if you mention how you installed numpy and PyTorch (for example what BLAS library are they using).

**Comment by apaszke** [comment]
Right now our GPU libs are very optimized, but we can't say the same about the CPU backend. It's getting better, but we still aren't there where numpy is, and this is probably why are you seeing this difference. Can you try running your script with `OMP_NUM_THREADS=X` where X is 8 (or less if you have fewer cores)?

**Comment by makeyourownneuralnetwork** [comment]
@pavanky I'm using the latest Anaconda Python 3.6 updated daily via conda

pytorch was installed using the conda method on the official pytorch GitHub readme

**Comment by makeyourownneuralnetwork** [comment]
@apaszke thanks for your helpful reply

My MacBook Pro 13 early 2015 has only 1 cpu with 2 cores.  I therefore set `OMP_NUM_THREADS=2 `and did the timings again.

The results were:

- PyTorch - **928 and 995** seconds compared to the 841 and 834 seconds earlier.
- Plain python with numpy again -  **591 and 597** seconds. 

 I did this  the comparison again as I am now running a browser showing live tv news, so need to make the environment equivalent to aid comparison. The ratio is **1.6** not in favour of PyTorch.

Interestingly, setting the `OMP_NUM_THREADS=1 `makes no difference, the result with PyTorch was **953** seconds, within the range above. 

The environment variable was set in the notebook using:
```
%env OMP_NUM_THREADS=1
```

and was tested with `%env`

**Comment by rosinality** [comment]
Autodifferentiaion engine of PyTorch is very fast (so performance is similar or even better than the other frameworks), but it can be slower than raw numpy forward/backward implementation, especially when network is small. I think performance will be better if you implement same network with torch.Tensor like numpy, instead of Variable.

**Comment by makeyourownneuralnetwork** [comment]
The readme.md on pytorch's GitHub page says:

> Hence, PyTorch is quite fast -- whether you run small or large neural networks.

But the above comments suggest this isn't true.

I am now setting up a Google Cloud Compute GPU instance to test performance with CUDA. Weill update this issue if interesting.

**Comment by makeyourownneuralnetwork** [comment]
I've updated my blog .. the GPU results are a bit faster .. about 25%.

```
CPU	GPU
	
494	366
483	372
451	355
	
476.0	364.3 (averages)
```

See here for more: [http://makeyourownneuralnetwork.blogspot.co.uk/2017/05/learning-mnist-with-gpu-acceleration.html](http://makeyourownneuralnetwork.blogspot.co.uk/2017/05/learning-mnist-with-gpu-acceleration.html)

I will do 2 more tests .. one using the environment variable ... and one scaling the hidden layer to about 1000, to see if the relative performance improves.

**Comment by makeyourownneuralnetwork** [comment]
Update - I did the tests comparing PyTorch in CPU vs GPU mode. The results are interesting and positive.

<img width="1033" alt="pytorch_cpu_v_gpu" src="https://cloud.githubusercontent.com/assets/17411198/26444967/448ed146-4136-11e7-8a45-6881db14e6fe.png">

In GPU mode, the duration to complete MNIST training rises slowly, but in CPU mode it grows quickly, as the size of the network grows (hidden nodes).

More detail at the bottom of this post: [http://makeyourownneuralnetwork.blogspot.co.uk/2017/05/learning-mnist-with-gpu-acceleration.html](http://makeyourownneuralnetwork.blogspot.co.uk/2017/05/learning-mnist-with-gpu-acceleration.html)

So this is good. I think PyTorch is nice to work with .. I. understand that the tutorials and beginner-friendly guides still need people to work on them... would be great when they're out. I do an intro talk at PyData monthly meetup in London on my experience beginning PyTorch. 

Now if only it was faster than numpy .....

**Comment by soumith** [comment]
great. I'd consider the issue closed then.
We will slowly and eventually converge to numpy for all of our workloads. Right now we are faster for some workloads on CPU, but as you showed slower for other.

**Comment by impredicative** [comment]
Why use `numpy.dot` when the `@` operator works?

**Comment by fmassa** [comment]
@impredicative yes, it supports the `@` operator, but it is equivalent to `matmul` and not `dot` (both also present in numpy)
